pkg/prometheus/server/statefulset.go

```diff
@@ -0,0 +1,771 @@
+			VolumeSource: v1.VolumeSource{
+				EmptyDir: &v1.EmptyDirVolumeSource{},
+			},
+		})
+	} else if storageSpec.EmptyDir != nil {
+		emptyDir := storageSpec.EmptyDir
+		statefulset.Spec.Template.Spec.Volumes = append(statefulset.Spec.Template.Spec.Volumes, v1.Volume{
+			Name: prompkg.VolumeName(objMeta.GetName()),
+			VolumeSource: v1.VolumeSource{
+				EmptyDir: emptyDir,
+			},
+		})
+	} else if storageSpec.Ephemeral != nil {
+		ephemeral := storageSpec.Ephemeral
+		statefulset.Spec.Template.Spec.Volumes = append(statefulset.Spec.Template.Spec.Volumes, v1.Volume{
+			Name: prompkg.VolumeName(objMeta.GetName()),
+			VolumeSource: v1.VolumeSource{
+				Ephemeral: ephemeral,
+			},
+		})
+	} else {
+		pvcTemplate := operator.MakeVolumeClaimTemplate(storageSpec.VolumeClaimTemplate)
+		if pvcTemplate.Name == "" {
+			pvcTemplate.Name = prompkg.VolumeName(objMeta.GetName())
+		}
+		if storageSpec.VolumeClaimTemplate.Spec.AccessModes == nil {
+			pvcTemplate.Spec.AccessModes = []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce}
+		} else {
+			pvcTemplate.Spec.AccessModes = storageSpec.VolumeClaimTemplate.Spec.AccessModes
+		}
+		pvcTemplate.Spec.Resources = storageSpec.VolumeClaimTemplate.Spec.Resources
+		pvcTemplate.Spec.Selector = storageSpec.VolumeClaimTemplate.Spec.Selector
+		statefulset.Spec.VolumeClaimTemplates = append(statefulset.Spec.VolumeClaimTemplates, *pvcTemplate)
+	}
+
+	statefulset.Spec.Template.Spec.Volumes = append(statefulset.Spec.Template.Spec.Volumes, cpf.Volumes...)
+
+	if cpf.HostNetwork {
+		statefulset.Spec.Template.Spec.DNSPolicy = v1.DNSClusterFirstWithHostNet
+	}
+
+	return statefulset, nil
+}
+
+func makeStatefulSetSpec(
+	logger log.Logger,
+	baseImage, tag, sha string,
+	retention monitoringv1.Duration,
+	retentionSize monitoringv1.ByteSize,
+	rules monitoringv1.Rules,
+	query *monitoringv1.QuerySpec,
+	allowOverlappingBlocks bool,
+	enableAdminAPI bool,
+	queryLogFile string,
+	thanos *monitoringv1.ThanosSpec,
+	disableCompaction bool,
+	p monitoringv1.PrometheusInterface,
+	c *operator.Config,
+	cg *prompkg.ConfigGenerator,
+	shard int32,
+	ruleConfigMapNames []string,
+	tlsAssetSecrets []string,
+) (*appsv1.StatefulSetSpec, error) {
+	// Prometheus may take quite long to shut down to checkpoint existing data.
+	// Allow up to 10 minutes for clean termination.
+	terminationGracePeriod := int64(600)
+	cpf := p.GetCommonPrometheusFields()
+	promName := p.GetObjectMeta().GetName()
+
+	pImagePath, err := operator.BuildImagePath(
+		operator.StringPtrValOrDefault(cpf.Image, ""),
+		operator.StringValOrDefault(baseImage, c.PrometheusDefaultBaseImage),
+		operator.StringValOrDefault(cpf.Version, operator.DefaultPrometheusVersion),
+		operator.StringValOrDefault(tag, ""),
+		operator.StringValOrDefault(sha, ""),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if cg.Version().Major != 2 {
+		return nil, errors.Errorf("unsupported Prometheus major version %s", cg.Version())
+	}
```

Comment:
Can't we move this `if` to `NewConfigGenerator`? From my POV the promcfg package should be responsible for these types of checks

Reply:
That makes perfect sense!